// mppx/dist/discovery/OpenApi.js
/**
 * Generates an OpenAPI 3.1.0 discovery document from an mppx instance
 * and route configuration.
 */
export function generate(mppx, config) {
    const methods = mppx.methods;
    const methodsByKey = new Map();
    const intentCount = {};
    for (const mi of methods) {
        methodsByKey.set(`${mi.name}/${mi.intent}`, mi);
        intentCount[mi.intent] = (intentCount[mi.intent] ?? 0) + 1;
    }
    for (const mi of methods) {
        if (intentCount[mi.intent] === 1)
            methodsByKey.set(mi.intent, mi);
    }
    const routes = config.routes.map((route) => resolveRoute(route, methodsByKey));
    return createDocument({
        info: {
            title: config.info?.title ?? mppx.realm,
            version: config.info?.version ?? '1.0.0',
        },
        routes,
        serviceInfo: config.serviceInfo,
    });
}
/**
 * Generates an OpenAPI 3.1.0 discovery document for a proxy surface.
 */
export function generateProxy(config) {
    const routes = config.routes.map((route) => ({
        ...route,
        path: withBasePath(config.basePath, route.path),
    }));
    return createDocument({
        info: {
            title: config.info?.title ?? 'API Proxy',
            version: config.info?.version ?? '1.0.0',
        },
        routes,
        serviceInfo: config.serviceInfo,
    });
}
function createDocument(config) {
    const paths = {};
    for (const route of config.routes) {
        const method = route.method.toLowerCase();
        const operation = {
            responses: {
                ...(route.payment ? { '402': { description: 'Payment Required' } } : {}),
                '200': { description: 'Successful response' },
            },
        };
        if (route.payment)
            operation['x-payment-info'] = route.payment;
        if (route.summary)
            operation.summary = route.summary;
        if (route.requestBody)
            operation.requestBody = route.requestBody;
        if (!paths[route.path])
            paths[route.path] = {};
        paths[route.path][method] = operation;
    }
    const doc = {
        info: config.info,
        openapi: '3.1.0',
        paths,
    };
    if (config.serviceInfo)
        doc['x-service-info'] = config.serviceInfo;
    return doc;
}
function resolveRoute(route, methodsByKey) {
    if ('handler' in route) {
        const internal = route.handler._internal;
        if (!internal)
            throw new Error(`Route ${route.method.toUpperCase()} ${route.path} is missing discovery metadata`);
        return {
            method: route.method,
            path: route.path,
            payment: paymentInfoFromCanonical({
                canonicalRequest: internal._canonicalRequest,
                intent: internal.intent,
                method: internal.name,
            }),
            ...(route.requestBody ? { requestBody: route.requestBody } : {}),
            ...(route.summary ? { summary: route.summary } : {}),
        };
    }
    const mi = methodsByKey.get(route.intent);
    if (!mi) {
        throw new Error(`Unknown intent "${route.intent}" for route ${route.method.toUpperCase()} ${route.path}. Available: ${[...methodsByKey.keys()].join(', ')}`);
    }
    return {
        method: route.method,
        path: route.path,
        payment: paymentInfoFromCanonical({
            canonicalRequest: route.options,
            intent: mi.intent,
            method: mi.name,
        }),
        ...(route.requestBody ? { requestBody: route.requestBody } : {}),
        ...(route.summary ? { summary: route.summary } : {}),
    };
}
function paymentInfoFromCanonical(route) {
    const { canonicalRequest, intent, method } = route;
    const methodDetails = (canonicalRequest.methodDetails ?? {});
    const amount = pickString(canonicalRequest.amount) ?? pickString(methodDetails.amount) ?? null;
    const currency = pickString(canonicalRequest.currency) ?? pickString(methodDetails.currency);
    const description = pickString(canonicalRequest.description);
    const base = {
        amount,
        ...(currency ? { currency } : {}),
        ...(description ? { description } : {}),
        intent,
        method,
    };
    // Forward any extra canonical params that aren't already covered.
    const reserved = new Set(['amount', 'currency', 'description', 'methodDetails']);
    for (const [key, value] of Object.entries(canonicalRequest)) {
        if (!reserved.has(key) && value !== undefined)
            base[key] = value;
    }
    return base;
}
function pickString(value) {
    return typeof value === 'string' ? value : undefined;
}
function withBasePath(basePath, path) {
    if (!basePath)
        return path;
    const normalizedBasePath = basePath.startsWith('/') ? basePath : `/${basePath}`;
    const trimmedBasePath = normalizedBasePath.endsWith('/')
        ? normalizedBasePath.slice(0, -1)
        : normalizedBasePath;
    return `${trimmedBasePath}${path}`;
}
