// mppx/dist/proxy/Proxy.js
import * as Credential from '../Credential.js';
import { generateProxy } from '../discovery/OpenApi.js';
import * as Request from '../server/Request.js';
import * as Headers from './internal/Headers.js';
import * as Route from './internal/Route.js';
import * as Service from './Service.js';
/**
 * Creates a paid API proxy.
 *
 * Routes incoming requests to upstream services, injects credentials,
 * and requires payment via the mppx 402 protocol for non-free endpoints.
 *
 * @example
 * ```ts
 * import { Proxy, openai } from 'mppx/proxy'
 * import { Mppx, tempo } from 'mppx/server'
 *
 * const mppx = Mppx.create({ methods: [tempo()] })
 *
 * const proxy = Proxy.create({
 *   services: [
 *     openai({
 *       apiKey: 'sk-...',
 *       routes: {
 *         'POST /v1/chat/completions': mppx.charge({ amount: '0.05' }),
 *         'GET /v1/models': true,
 *       },
 *     }),
 *   ],
 * })
 * ```
 */
export function create(config) {
    const fetchImpl = config.fetch ?? globalThis.fetch;
    const services = new Map(config.services.map((s) => {
        const proxy = createFetchProxy(s.baseUrl, { fetch: fetchImpl });
        return [s.id, { service: s, proxy }];
    }));
    // Pre-generate static discovery responses once at startup.
    const openApiJson = JSON.stringify(generateProxy({
        basePath: config.basePath,
        info: {
            title: config.title ?? 'API Proxy',
            version: config.version ?? '1.0.0',
        },
        routes: buildDiscoveryRoutes(config.services),
        serviceInfo: buildServiceInfo(config),
    }));
    const llmsTxt = Service.toLlmsTxt(config.services, {
        title: config.title,
        description: config.description,
        openApiPath: withBasePath(config.basePath, '/openapi.json'),
    });
    async function handle(request) {
        const url = new URL(request.url);
        const pathname = Route.pathname(url, config.basePath);
        if (!pathname)
            return new Response('Not Found', { status: 404 });
        if (request.method === 'GET' &&
            (pathname === '/openapi.json' || pathname === '/openapi.json/')) {
            return new Response(openApiJson, {
                headers: {
                    'Cache-Control': 'public, max-age=300',
                    'Content-Type': 'application/json',
                },
            });
        }
        if (request.method === 'GET' && pathname === '/llms.txt')
            return new Response(llmsTxt, {
                headers: { 'Content-Type': 'text/plain; charset=utf-8' },
            });
        const parsed = Route.parse(pathname);
        if (!parsed)
            return new Response('Not Found', { status: 404 });
        const { serviceId, upstreamPath } = parsed;
        const entry = services.get(serviceId);
        if (!entry)
            return new Response('Not Found', { status: 404 });
        const { service, proxy } = entry;
        const exactMatch = Route.match(service.routes, request.method, upstreamPath);
        const fallbackBinding = !exactMatch && request.method === 'POST' && request.headers.has('authorization')
            ? getPaymentBinding(request)
            : null;
        const fallbackMatch = !exactMatch && request.method === 'POST' && request.headers.has('authorization')
            ? // Management POSTs (e.g. session close) may target a path whose route
                // is registered for a different HTTP method (e.g. GET). Fall back to
                // path-only matching so the payment handler can process the action.
                // When the credential parses cleanly, also bind on payment method+intent
                // so same-path paid routes can coexist without sharing credentials.
                Route.matchPath(service.routes, upstreamPath, 
                // skip free routes (e.g. `'GET /foo/bar': true`)
                (endpoint) => endpoint !== true && matchesPaymentBinding(endpoint, fallbackBinding))
            : null;
        const matched = exactMatch ?? fallbackMatch;
        if (!matched)
            return new Response('Not Found', { status: 404 });
        const endpoint = matched.value;
        const ctx = { request, service, upstreamPath };
        if (endpoint === true)
            return proxyUpstream({ request, service, ctx, proxy });
        const handler = typeof endpoint === 'function' ? endpoint : endpoint.pay;
        const result = await handler(request);
        if (result.status === 402)
            return result.challenge;
        const managementResponse = (() => {
            try {
                return result.withReceipt();
            }
            catch (error) {
                if (error instanceof Error &&
                    error.message === 'withReceipt() requires a response argument')
                    return null;
                throw error;
            }
        })();
        if (managementResponse)
            return managementResponse;
        if (fallbackMatch)
            return new Response('Method Not Allowed', { status: 405 });
        const options = Service.getOptions(endpoint);
        const upstreamRes = await proxyUpstream({
            request,
            service,
            ctx: { ...ctx, ...options },
            proxy,
        });
        return result.withReceipt(upstreamRes);
    }
    return {
        fetch: handle,
        listener: Request.toNodeListener(handle),
    };
}
/** @internal */
async function proxyUpstream(options) {
    const { request, service, ctx, proxy } = options;
    const url = ctx.upstreamPath + new URL(request.url).search;
    const headers = Headers.scrub(request.headers);
    const method = request.method.toUpperCase();
    const hasBody = method !== 'GET' && method !== 'HEAD';
    const init = {
        method: request.method,
        headers,
        signal: request.signal,
    };
    if (hasBody && request.body) {
        init.body = request.body;
        init.duplex = 'half';
    }
    let upstreamReq = new globalThis.Request(new URL(url, new URL(service.baseUrl).origin), init);
    if (service.rewriteRequest)
        upstreamReq = await service.rewriteRequest(upstreamReq, ctx);
    let upstreamRes = await proxy(upstreamReq);
    upstreamRes = Headers.scrubResponse(upstreamRes);
    if (service.rewriteResponse)
        upstreamRes = await service.rewriteResponse(upstreamRes, ctx);
    return upstreamRes;
}
function buildDiscoveryRoutes(services) {
    return services.flatMap((service) => Object.entries(service.routes).map(([pattern, endpoint]) => {
        const tokens = pattern.trim().split(/\s+/);
        const hasMethod = tokens.length >= 2;
        const path = hasMethod ? tokens.slice(1).join(' ') : tokens[0];
        return {
            method: hasMethod ? tokens[0] : 'GET',
            path: `/${service.id}${path}`,
            payment: endpoint ? Service.paymentOf(endpoint) : null,
        };
    }));
}
function buildServiceInfo(config) {
    const categories = config.categories ??
        Array.from(new Set(config.services.flatMap((service) => service.categories ?? [])));
    const docs = {
        ...(config.docs ?? {}),
        llms: config.docs?.llms ?? withBasePath(config.basePath, '/llms.txt'),
    };
    return {
        ...(categories.length > 0 ? { categories } : {}),
        docs,
    };
}
function withBasePath(basePath, path) {
    if (!basePath)
        return path;
    const normalized = basePath.startsWith('/') ? basePath : `/${basePath}`;
    const trimmed = normalized.endsWith('/') ? normalized.slice(0, -1) : normalized;
    return `${trimmed}${path}`;
}
function getPaymentBinding(request) {
    try {
        const credential = Credential.fromRequest(request);
        return {
            intent: credential.challenge.intent,
            method: credential.challenge.method,
        };
    }
    catch {
        return null;
    }
}
function matchesPaymentBinding(endpoint, binding) {
    if (endpoint === true)
        return false;
    if (!binding)
        return true;
    const payment = Service.paymentOf(endpoint);
    if (!payment)
        return true;
    return payment.method === binding.method && payment.intent === binding.intent;
}
function createFetchProxy(target, options) {
    const localFetch = options?.fetch ?? globalThis.fetch;
    const targetUrl = new URL(target);
    if (targetUrl.pathname.endsWith('/'))
        targetUrl.pathname = targetUrl.pathname.replace(/\/+$/, '');
    return async (input, init) => {
        const request = new globalThis.Request(input, init);
        const url = new URL(request.url);
        const proxyUrl = new URL(url.search, targetUrl);
        if (url.pathname !== '/')
            proxyUrl.pathname =
                proxyUrl.pathname === '/' ? url.pathname : proxyUrl.pathname + url.pathname;
        const proxyInit = {
            method: request.method,
            headers: new globalThis.Headers(request.headers),
            signal: request.signal,
            redirect: request.redirect,
            ...init,
        };
        if (request.method !== 'GET' && request.method !== 'HEAD') {
            proxyInit.body = request.body;
            proxyInit.duplex = 'half';
        }
        return localFetch(proxyUrl, proxyInit);
    };
}
