// incur/dist/Openapi.js
import { z } from 'zod';
import * as Fetch from './Fetch.js';
import { dereference } from './internal/dereference.js';
/** Generates incur command entries from an OpenAPI spec. Resolves all `$ref` pointers. */
export async function generateCommands(spec, fetch, options = {}) {
    const resolved = dereference(structuredClone(spec));
    const commands = new Map();
    const paths = (resolved.paths ?? {});
    for (const [path, methods] of Object.entries(paths)) {
        for (const [method, operation] of Object.entries(methods)) {
            if (method.startsWith('x-'))
                continue;
            const op = operation;
            const name = op.operationId ?? `${method}_${path.replace(/[/{}]/g, '_')}`;
            const httpMethod = method.toUpperCase();
            const pathParams = (op.parameters ?? []).filter((p) => p.in === 'path');
            const queryParams = (op.parameters ?? []).filter((p) => p.in === 'query');
            const bodySchema = op.requestBody?.content?.['application/json']?.schema;
            const bodyProps = (bodySchema?.properties ?? {});
            const bodyRequired = new Set(bodySchema?.required ?? []);
            // Build args Zod schema from path params
            let argsSchema;
            if (pathParams.length > 0) {
                const shape = {};
                for (const p of pathParams) {
                    let zodType = p.schema ? toZod(p.schema) : z.string();
                    if (p.description)
                        zodType = zodType.describe(p.description);
                    // Path params need coercion from string argv
                    shape[p.name] = coerceIfNeeded(zodType);
                }
                argsSchema = z.object(shape);
            }
            // Build options Zod schema from query params + body properties
            const optShape = {};
            for (const p of queryParams) {
                let zodType = p.schema ? toZod(p.schema) : z.string();
                if (!p.required)
                    zodType = zodType.optional();
                if (p.description)
                    zodType = zodType.describe(p.description);
                optShape[p.name] = coerceIfNeeded(zodType);
            }
            for (const [key, schema] of Object.entries(bodyProps)) {
                let zodType = toZod(schema);
                if (!bodyRequired.has(key))
                    zodType = zodType.optional();
                optShape[key] = zodType;
            }
            const optionsSchema = Object.keys(optShape).length > 0 ? z.object(optShape) : undefined;
            commands.set(name, {
                description: op.summary ?? op.description,
                args: argsSchema,
                options: optionsSchema,
                run: createHandler({
                    basePath: options.basePath,
                    fetch,
                    httpMethod,
                    path,
                    pathParams,
                    queryParams,
                    bodyProps,
                }),
            });
        }
    }
    return commands;
}
function createHandler(config) {
    return async (context) => {
        const { args = {}, options = {} } = context;
        // Build URL path with interpolated path params
        let urlPath = (config.basePath ?? '') + config.path;
        for (const p of config.pathParams) {
            const value = args[p.name];
            if (value !== undefined)
                urlPath = urlPath.replace(`{${p.name}}`, String(value));
        }
        // Build query string from query params
        const query = new URLSearchParams();
        for (const p of config.queryParams) {
            const value = options[p.name];
            if (value !== undefined)
                query.set(p.name, String(value));
        }
        // Build body from body properties
        let body;
        const bodyKeys = Object.keys(config.bodyProps);
        if (bodyKeys.length > 0) {
            const bodyObj = {};
            for (const key of bodyKeys)
                if (options[key] !== undefined)
                    bodyObj[key] = options[key];
            if (Object.keys(bodyObj).length > 0)
                body = JSON.stringify(bodyObj);
        }
        const input = {
            path: urlPath,
            method: config.httpMethod,
            headers: new Headers(),
            body,
            query,
        };
        if (body)
            input.headers.set('content-type', 'application/json');
        const request = Fetch.buildRequest(input);
        const response = await config.fetch(request);
        const output = await Fetch.parseResponse(response);
        if (!output.ok)
            return context.error({
                code: `HTTP_${output.status}`,
                message: typeof output.data === 'object' && output.data !== null && 'message' in output.data
                    ? String(output.data.message)
                    : typeof output.data === 'string'
                        ? output.data
                        : `HTTP ${output.status}`,
            });
        return output.data;
    };
}
/** Converts a JSON Schema object to a Zod schema. */
function toZod(schema) {
    return z.fromJSONSchema(schema);
}
/** Wraps a Zod schema with coercion if the base type is number or boolean (argv is always strings). */
function coerceIfNeeded(schema) {
    const isOptional = schema instanceof z.ZodOptional;
    const inner = isOptional ? schema.unwrap() : schema;
    const coerced = (() => {
        // Direct number
        if (inner instanceof z.ZodNumber)
            return isOptional ? z.coerce.number().optional() : z.coerce.number();
        // Direct boolean
        if (inner instanceof z.ZodBoolean)
            return isOptional ? z.coerce.boolean().optional() : z.coerce.boolean();
        // Union containing number or boolean (e.g. type: ["number", "null"] from OpenAPI 3.1)
        if (inner instanceof z.ZodUnion) {
            const options = inner._zod?.def?.options;
            if (options?.some((o) => o instanceof z.ZodNumber))
                return isOptional ? z.coerce.number().optional() : z.coerce.number();
            if (options?.some((o) => o instanceof z.ZodBoolean))
                return isOptional ? z.coerce.boolean().optional() : z.coerce.boolean();
        }
        // No coercion needed
        return undefined;
    })();
    if (!coerced)
        return schema;
    const desc = schema.description ?? inner.description;
    return desc ? coerced.describe(desc) : coerced;
}
