// mppx/dist/tempo/Attribution.js
import { Bytes, Hash, Hex } from 'ox';
/**
 * MPP attribution memo encoding for TIP-20 `transferWithMemo`.
 *
 * When no user-provided memo is present, the SDK auto-generates an
 * attribution memo so MPP transactions are identifiable on-chain.
 *
 * ## Byte Layout (32 bytes)
 *
 * | Offset | Size | Field                                     |
 * |--------|------|-------------------------------------------|
 * | 0..3   | 4    | TAG = keccak256("mpp")[0..3]               |
 * | 4      | 1    | version (0x01)                            |
 * | 5..14  | 10   | serverId = keccak256(serverId)[0..9]       |
 * | 15..24 | 10   | clientId = keccak256(clientId)[0..9] or 0s |
 * | 25..31 | 7    | nonce = keccak256(challengeId)[0..6]      |
 *
 * The TAG prefix makes MPP transactions trivially distinguishable
 * from arbitrary memos via `TransferWithMemo` event topic filtering.
 *
 * @module
 */
/** First 4 bytes of keccak256("mpp") — the on-chain MPP tag. */
export const tag = Hex.slice(Hash.keccak256(Bytes.fromString('mpp'), { as: 'Hex' }), 0, 4);
/** Current memo version. */
const version = 0x01;
/** 10 zero bytes representing an anonymous (no clientId) client. */
export const anonymous = '0x00000000000000000000';
/**
 * Computes a 10-byte fingerprint from a string via keccak256.
 * @internal
 */
function fingerprint(value) {
    const hash = Hash.keccak256(Bytes.fromString(value), { as: 'Hex' });
    return Hex.toBytes(Hex.slice(hash, 0, 10));
}
/**
 * Encodes an MPP attribution memo as a `bytes32` hex string.
 *
 * @param parameters - The serverId (server identity) and optional clientId.
 * @returns A `0x`-prefixed 64-char hex string (32 bytes).
 *
 * @example
 * ```ts
 * import * as Attribution from './Attribution.js'
 *
 * const memo = Attribution.encode({ challengeId: 'challenge-123', clientId: 'my-app', serverId: 'api.example.com' })
 * ```
 */
export function encode(parameters) {
    const { serverId, clientId, challengeId } = parameters;
    const buf = new Uint8Array(32);
    buf.set(Hex.toBytes(tag), 0);
    buf[4] = version;
    buf.set(fingerprint(serverId), 5);
    if (clientId)
        buf.set(fingerprint(clientId), 15);
    // Derive the nonce from keccak256(challengeId)[0..6] to bind the memo
    // to the challenge and prevent transaction hash stealing.
    // TODO: expand to full memo verification once the server tracks the complete attribution payload.
    buf.set(challengeNonce(challengeId), 25);
    return Hex.fromBytes(buf);
}
/**
 * Checks whether a memo was generated by the MPP attribution system.
 *
 * @param memo - A `0x`-prefixed hex string (bytes32).
 * @returns `true` if the memo starts with the MPP tag and version byte.
 *
 * @example
 * ```ts
 * import * as Attribution from './Attribution.js'
 *
 * Attribution.isMppMemo(Attribution.encode({ challengeId: 'challenge-123', serverId: 'api.example.com' })) // true
 * Attribution.isMppMemo('0x0000...0000')      // false
 * ```
 */
export function isMppMemo(memo) {
    if (memo.length !== 66)
        return false;
    const memoTag = memo.slice(0, 10);
    const memoVersion = Number.parseInt(memo.slice(10, 12), 16);
    return memoTag.toLowerCase() === tag.toLowerCase() && memoVersion === version;
}
/**
 * Verifies that a memo's server fingerprint matches the given serverId.
 *
 * Checks TAG, version byte, and that bytes 5–14 equal keccak256(serverId)[0..9].
 *
 * @param memo - A `0x`-prefixed hex string (bytes32).
 * @param serverId - The expected server identity.
 * @returns `true` if the memo has a valid MPP tag and matching server fingerprint.
 */
export function verifyServer(memo, serverId) {
    if (!isMppMemo(memo))
        return false;
    const memoServerHex = `0x${memo.slice(12, 32)}`;
    const expectedHex = Hex.fromBytes(fingerprint(serverId));
    return memoServerHex.toLowerCase() === expectedHex.toLowerCase();
}
/**
 * Decodes an MPP attribution memo into its constituent parts.
 *
 * @param memo - A `0x`-prefixed hex string (bytes32).
 * @returns The decoded fields, or `null` if the memo is not an MPP memo.
 *
 * @example
 * ```ts
 * import * as Attribution from './Attribution.js'
 *
 * const memo = Attribution.encode({ challengeId: 'challenge-123', clientId: 'my-app', serverId: 'api.example.com' })
 * const decoded = Attribution.decode(memo)
 * // { version: 1, serverFingerprint: '0x...', clientFingerprint: '0x...', nonce: '0x...' }
 * ```
 */
export function decode(memo) {
    if (!isMppMemo(memo))
        return null;
    const memoVersion = Number.parseInt(memo.slice(10, 12), 16);
    const serverFingerprint = `0x${memo.slice(12, 32)}`;
    const clientHex = `0x${memo.slice(32, 52)}`;
    const nonce = `0x${memo.slice(52)}`;
    const clientFingerprint = clientHex.toLowerCase() === anonymous.toLowerCase() ? null : clientHex;
    return { version: memoVersion, serverFingerprint, clientFingerprint, nonce };
}
/**
 * Computes the 7-byte challenge-bound nonce: keccak256(challengeId)[0..6].
 * @internal
 */
export function challengeNonce(challengeId) {
    const hash = Hash.keccak256(Bytes.fromString(challengeId), { as: 'Hex' });
    return Hex.toBytes(Hex.slice(hash, 0, 7));
}
/**
 * Verifies that a memo's nonce is bound to the given challengeId.
 *
 * Checks TAG, version byte, and that bytes 25–31 equal keccak256(challengeId)[0..6].
 *
 * @param memo - A `0x`-prefixed hex string (bytes32).
 * @param challengeId - The expected challenge ID.
 * @returns `true` if the memo has a valid MPP tag and matching challenge nonce.
 */
export function verifyChallengeBinding(memo, challengeId) {
    const decoded = decode(memo);
    if (!decoded)
        return false;
    const expectedNonce = Hex.fromBytes(challengeNonce(challengeId));
    return decoded.nonce.toLowerCase() === expectedNonce.toLowerCase();
}
